rules/opentitan.bzl

```diff
@@ -726,12 +771,107 @@ pick_correct_archive_for_device = rv_rule(
 )
 
+def opentitan_multislot_flash_binary(
+        name,
+        srcs,
+        image_size,
+        key_per_binary,
+        offset_per_binary,
+        platform = OPENTITAN_PLATFORM):
+    """A helper macro for generating multislot OpenTitan binary flash images.
+
+    This macro is mostly a wrapper around the `assemble_flash_image` rule, that
+    invokes `opentitantool` to stitch together multiple `opentitan_flash_binary`
+    images to create a single image for bootstrapping. This enables efficient
+    testing by only requiring one boostrap operation to load both silicon
+    creator and owner stages of flash.
+    Args:
+      @param name: The name of this rule.
+      @param srcs: A list of `opentitan_flash_binary` targets to stich together.
+      @param image_size: The final flash image_size to pass to `opentitantool`.
+      @param key_per_binary: A dictionary mapping `opentitan_flash_binary`
+                             targets to the signing key names that were used for
+                             signing each target.
+      @param offset_per_binary: A dictionary mapping `opentitan_flash_binary`
+                                targets to address offsets that each image
+                                should be placed at.
+      @param platform: The target platform for the artifacts.
+    Emits rules:
+      For each device in per_device_deps entry:
+        rules emitted by `opentitan_binary` named: see `opentitan_binary` macro
+        assemble_flash_image named: <name>_<device>_bin_signed
+        bin_to_vmem          named: <name>_<device>_vmem64_signed
+        scrambled_flash_vmem named: <name>_<device>_scr_vmem64_signed
+        filegroup            named: <name>_<device>
+          Containing all targets for a single device for the above generated rules.
+        filegroup            named: <name>
+          Containing all targets across all devices for the above generated rules.
+    """
+    all_targets = []
+    for (device, _) in PER_DEVICE_DEPS.items():
+        devname = "{}_{}".format(name, device)
+        dev_targets = []
+        signed_dev_binaries = {}
+        for src in srcs:
+            if src not in key_per_binary:
+                fail("Missing signing key for binary: {}".format(src))
+            if src not in offset_per_binary:
+                fail("Missing offset for binary: {}".format(src))
+            key = key_per_binary[src]
+            offset = offset_per_binary[src]
+            signed_dev_binary = "{}_{}_bin_signed_{}".format(
+                src,
+                device,
+                key,
+            )
+            signed_dev_binaries[signed_dev_binary] = offset
+
+        # Assemble the signed binaries into a single binary.
+        signed_bin_name = "{}_bin_signed".format(devname)
+        dev_targets.append(":" + signed_bin_name)
+        assemble_flash_image(
+            name = signed_bin_name,
+            output = "{}.signed.bin".format(devname),
```

Comment:
nit: should the name of the output include offset/key names/etc or would it be too long?

Reply:
it can, it's up to the developer that uses the macro to name the output file however they would like. If the name passed the macro is `foo_slot_a_key_0_bar_slot_b_key_1`, the ouput file names are `foo_slot_a_key_0_bar_slot_b_key_1.signed.bin`. I didn't want to force a naming scheme here, in case developers wanted to use a generic name to refer to a specific multislot image.